src/MonteCarlo/WangLandau.jl

```diff
@@ -1,25 +1,39 @@
-# data for multicanonical ensemble -- either MUCA or WL
-mutable struct WangLandau{PR}
+mutable struct WangLandau{N, PR}
+    sys::System{N}
     ln_g::BinnedArray{Float64, Float64} # log density of states, ln[g(E)]
     hist::BinnedArray{Float64, Float64} # energy histogram
-    bounds::Vector{Float64} # energy bounds to sample [E_min, E_max]
+    bounds::Tuple{Float64, Float64} # energy bounds to sample [E_min, E_max]
     propose::PR     # MC move proposal (function)
-    per_spin::Bool  # choose whether to use energy per spin 
-    ln_f::Float64   # factor for insstantaneous modification of ln[g(E)] 
-    E::Float64      # current energy state
-
-    function WangLandau(; bin_size=1.0, bounds=Float64[], propose=propose_uniform, per_spin=true, ln_f=0.0)
-        new{typeof(propose)}(
-            BinnedArray{Float64, Float64}(bin_size=bin_size), 
-            BinnedArray{Float64, Float64}(bin_size=bin_size), 
-            bounds, propose, per_spin, ln_f, 0.0
+    ln_f::Float64   # visiting a site with energy E sets: ln[g(E)] += ln_f
+    E::Float64      # current energy state (per spin)
+
+    function WangLandau(; sys::System{N}, bin_size, bounds, propose::PR, ln_f=1.0, max_sweeps_relax=100) where {N, PR}
+        relax_system_to_bounds!(sys, bounds, propose, max_sweeps_relax)
+        E = energy(sys)/length(all_sites(sys))
+        return new{N, PR}(
+            sys,
+            BinnedArray{Float64, Float64}(; bin_size),
+            BinnedArray{Float64, Float64}(; bin_size),
+            bounds, propose, ln_f, E
         )
     end
 end
 
-# copy constructor
-function Base.copy(WL::WangLandau{PR}) where{PR}
-    return WangLandau(; WL.hist.bin_size, WL.bounds, WL.propose, WL.per_spin, WL.ln_f)
+# use MCMC sampling to get system into bounded energy range
+function relax_system_to_bounds!(sys::System{N}, bounds, propose::PR, max_sweeps) where {N, PR}
+    nsites = length(all_sites(sys))
+
+    E = energy(sys)/nsites
+    inbounds(E, bounds) && return
+    kT = E < bounds[1] ? Inf : 0.0
```

Comment:
I think zero temperature will work. That is, `1/0 == Inf` and `exp(-Inf) ==1`. Although as you point out, it may be faster to include a conditional there.